### PR TITLE
1271 utils use jest framwork only

### PR DESCRIPTION
**Pull request by billli0**
### Issue #:
[1271- Use a single test framework : **Utils module only**](https://github.com/beyondessential/tupaia-backlog/issues/1271)

This branch is a draft PR for tracking the changes.

### CODE REVIEW

File: packages/utils/scripts/JestTransformer.js [hunk 1]
```diff
@@ -0,0 +1,10 @@
+/**
+ * https://babeljs.io/docs/en/config-files#jest
+ *
+ * Setup jest in Monorepo. This is to create a custom jest transformer file that wraps babel-jest.
+ *
+ * @type {Transformer}
+ */
+module.exports = require('babel-jest').createTransformer({
+  rootMode: 'upward',
+});
```
Comment (kael89) [inline packages/utils/scripts/JestTransformer.js:10]: Nice solution!! We'll probably want to move this in the root folder in case it's also required in other packages, but don't worry about this for now

File: packages/utils/src/__tests__/WorkBookParser.test.js [hunk 2]
```diff
@@ -3,14 +3,12 @@
...
 
-import chai, { expect } from 'chai';
-import chaiAsPromised from 'chai-as-promised';
 import xlsx from 'xlsx';
 import sinon from 'sinon';
 
 import { WorkBookParser } from '../WorkBookParser';
 
-chai.use(chaiAsPromised);
+// chai.use(chaiAsPromised);
```
Comment (kael89) [inline packages/utils/src/__tests__/WorkBookParser.test.js:11]: Let's remove this comment

```suggestion
```

File: packages/utils/src/__tests__/checkValueSatisfiesCondition.test.js [hunk 3]
```diff
@@ -1,114 +1,111 @@
...
+      expect(() => checkValueSatisfiesCondition('hello', { operator: 'NOT_VALID' })).toThrowError("Unknown operator: 'NOT_VALID'");
     });
 
     it('should correctly check if a value satisfies a plain condition', () => {
-      expect(checkValueSatisfiesCondition('hello', 'hello')).to.be.true;
-      expect(checkValueSatisfiesCondition(0, 0)).to.be.true;
-      expect(checkValueSatisfiesCondition(true, true)).to.be.true;
-      expect(checkValueSatisfiesCondition(false, false)).to.be.true;
-      expect(checkValueSatisfiesCondition(null, null)).to.be.true;
+      expect(checkValueSatisfiesCondition('hello', 'hello')).toBe(true);
```
Comment (kael89) [inline packages/utils/src/__tests__/checkValueSatisfiesCondition.test.js]: This file seems like a great candidate for using parameterized tests. Let's do this in another PR though!

File: packages/utils/src/__tests__/object.test.js [hunk 4]
```diff
@@ -31,49 +29,50 @@ const object3 = {
 const assertSortingCorrectness = (sortingMethod, input, expectedValue) => {
   const arrayToSort = [...input];
   arrayToSort.sort(sortingMethod);
-  expect(arrayToSort).to.deep.equal(expectedValue);
+  expect(arrayToSort).toEqual(expectedValue);
```
Comment (kael89) [inline packages/utils/src/__tests__/object.test.js:32]: Although `toEqual` is possibly fine for what we want to achieve here, let's use the stricter version ([toStrictEqual](https://jestjs.io/docs/en/expect#tostrictequalvalue)]), unless if we have specific reasons to use `toEqual`

Similarly for all other cases in this PR

```suggestion
  expect(arrayToSort).toStrictEqual(expectedValue);
```

File: packages/utils/scripts/JestTransformer.js (hunk 1 above)
Comment (billli0) [reply packages/utils/scripts/JestTransformer.js:10]: Good idea!

File: packages/utils/src/__tests__/datetime.test.js [hunk 5]
```diff
@@ -21,7 +20,8 @@ describe('DateTime', () => {
       'Antarctica/McMurdo', // +13 while in DST time.
     ];
 
-    expect(moment.parseZone(timestamp).format('Z')).to.equal(utcOffset);
-    expect(getTimezoneNameFromTimestamp(timestamp)).to.be.oneOf(timezones);
+    expect(moment.parseZone(timestamp).format('Z')).toBe(utcOffset);
+    // expect(getTimezoneNameFromTimestamp(timestamp)).to.be.oneOf(timezones);
```
Comment (kael89) [inline packages/utils/src/__tests__/datetime.test.js:24]:
```suggestion
```

File: packages/utils/src/__tests__/checkValueSatisfiesCondition.test.js [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+        "Unknown operator: 'NOT_VALID'",
+      );
+    });
+
+    it.each([
+      ['hello', 'hello', true],
+      [0, 0, true],
+      [true, true, true],
+      [false, false, true],
+      [null, null, true],
```
Comment (kael89) [inline packages/utils/src/__tests__/checkValueSatisfiesCondition.test.js:16]: Nice conversion to parameterized tests in this file!!

I think it's a good call to not assign to `testData` here, but rather provide them as a direct input in `it.each` 💯

File: packages/utils/src/__tests__/WorkBookParser.test.js [hunk 7]
```diff
@@ -55,9 +58,10 @@ describe('WorkBookParser', () => {
       });
     });
 
-    it('should map the input to itself if no mapper is provided', () => {
+    it('should map the input to itself if no mapper is provided', async () => {
       const parser = new WorkBookParser();
-      return expect(parser.parse(WORK_BOOK)).to.eventually.deep.equal(SHEETS);
+      const result = await parser.parse(WORK_BOOK);
+      return expect(result).toEqual(SHEETS);
```
Comment (kael89) [inline packages/utils/src/__tests__/WorkBookParser.test.js:64]: Although the above syntax is totally fine, let's try to convert it to:

```suggestion
      return expect(parser.parse(WORK_BOOK)).resolves.toStictEqual(SHEETS);
```

since it is more similar to the non asynx `expect` syntax, retaining an input => output flow

Similarly for the other test cases in this file where we `await` results and then `expect` on them

File: packages/utils/src/__tests__/array.test.js [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+      [undefined, null, []].forEach(input => expect(min(input)).toBeUndefined());
+    });
+  });
+
+  describe('max', () => {
+    const testData = [
+      [[2, 3, 10], 10],
+      [[-20, 1, 3, 10], 10],
+    ];
+    it.each(testData)(
```
Comment (kael89) [inline packages/utils/src/__tests__/array.test.js:64]: I believe the pattern you established in `checkValueSatisfiesCondition.test.js` is also appropriate here:

```suggestion
    it.each(testData)([
      [[2, 3, 10], 10],
      [[-20, 1, 3, 10], 10],
    ])(
```

File: packages/utils/src/__tests__/array.test.js [hunk 9]
```diff
@@ -0,0 +1,75 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
+ */
+
+import { countDistinct, min, max } from '../array';
+
+describe('array', () => {
+  const testData = [
+    ['countDistinct', [], 0],
```
Comment (kael89) [inline packages/utils/src/__tests__/array.test.js:10]:
```suggestion
    ['empty array', [], 0],
```

File: packages/utils/src/__tests__/array.test.js [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    ['single item ', [0], 1],
+    ['multiple items, same value', [0], 1],
+    ['multiple items, different values', [0, 1], 2],
+    ['multiple items, mixed same & different values', [0, 1, 0, 1, 1, 2], 3],
+  ];
+  describe('countDistinct', () => {
+    it.each(testData)('%s', (string, input, expected) => {
```
Comment (kael89) [inline packages/utils/src/__tests__/array.test.js:17]: Let's also pass the `testData` from above directly here

File: packages/utils/src/__tests__/array.test.js [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+      expect(countDistinct(input)).toBe(expected);
+    });
+
+    describe('custom mappers', () => {
+      const customMappersTestData = [
```
Comment (kael89) [inline packages/utils/src/__tests__/array.test.js:22]: And then this can become

```suggestion
      const testData = [
```

**Review by kael89 — CHANGES_REQUESTED** [review]
Excellent work!! I love how you paid attention to small details, and devised solutions to problems we haven't solved before 🎉 🥇 

All my suggestions are about details

File: packages/utils/src/__tests__/checkValueSatisfiesCondition.test.js (hunk 3 above)
Comment (kael89) [reply packages/utils/src/__tests__/checkValueSatisfiesCondition.test.js]: You've already done it! 🙌

File: packages/utils/src/__tests__/array.test.js (hunk 10 above)
Comment (billli0) [reply packages/utils/src/__tests__/array.test.js:17]: > Let's also pass the `testData` from above directly here

Good idea Kostas. I'll add it in the next commit.

**Review by kael89 — CHANGES_REQUESTED** [review]
Almost there! Could you please also convert all `.toEqual` in this package to `.toStrictEqual`

**Comment by billli0** [comment]
> Almost there! Could you please also convert all `.toEqual` in this package to `.toStrictEqual`

Yes I should put that in mind. Thanks for checking Kostas.

**Review by kael89 — APPROVED** [review]
Amazing work! Approved, but please hold off merging until I push some changes regarding the `skipped` tests

**Comment by billli0** [comment]
> Amazing work! Approved, but please hold off merging until I push some changes regarding the `skipped` tests

Yes sure thing! Can't wait to see your work!

**Comment by billli0** [comment]
This morning I've made a script to work for converting to parameterized tests. Even though it still requires some manual fix, it increases efficency and boosts my mood 👍

**Comment by kael89** [comment]
> This morning I made a script for converting tests to parameterized tests. Even though it still requires some manual fix, it increases efficency and boosts my mood 👍

That's great! It makes sense to commit this script in our source code, after it's finalised 🎉  👌

**Comment by billli0** [comment]
> > This morning I made a script for converting tests to parameterized tests. Even though it still requires some manual fix, it increases efficency and boosts my mood 👍
> 
> That's great! It makes sense to commit this script in our source code, after it's finalised 🎉 👌

definitely:raised_hands:

**Review by kael89 — APPROVED** [review]

**Comment by kael89** [comment]
@billli0 This can be merged straight into `dev` (without QA) since it is developer-related - no production code was touched
